README.md

```diff
@@ -584,44 +600,44 @@ def build(env):
     # and use this information for a new template.
     env.template("module_header.hpp.in", substitutions={"headers": headers})
 
-    # You can add metadata to the build log which then made
-    # available in the post_build step. This is like a dictionary.
-    env.add_metadata("include_path", "generated_folder/")
+    # Add values to a collector, all these are type checked
+    env.collect("::include_path", "repo/must_be_valid_path/", "repo/folder2/")
 
 
 # The post build step can do everything the build step can,
 # but you can't add to the metadata anymore:
-# - env.add_metadata() unavailable
+# - env.collect() unavailable
 # You have access to the entire buildlog up to this point
-def post_build(env, buildlog):
+def post_build(env):
     # The absolute path to the lbuild output directory
-    outpath = buildlog.outpath
+    outpath = env.buildlog.outpath
 
     # All modules that were built
-    modules = buildlog.modules
+    modules = env.buildlog.modules
     # All file generation operations that were done
-    operations = buildlog.operations
+    operations = env.buildlog.operations
     # All operations per module
-    operations = buildlog.operations_per_module("repo:module")
+    operations = env.buildlog.operations_per_module("repo:module")
 
     # iterate over all operations directly
     for operation in buildlog:
         # Get the module name that generated this file
-        env.log.info("{} generated the '{}' file".format(
-                     operation.module_name, operation.filename_out()))
+        env.log.info("Module({}) generated the '{}' file"
+                     .format(operation.module, operation.filename))
         # You can also get the filename relative to a subfolder in outpath
-        operation.filename_out(path="subfolder/")
+        env.relative_output(operation.filename, "subfolder/")
+        # or as an absolute path
+        env.real_output(operation.filename, "subfolder/")
 
-    # get the metadata: this is a dictionary of lists!
-    metadata = buildlog.metadata
-    include_paths = [("-I" + p) for p in metadata.get("include_path", [])]
+    # get all include paths from all active modules
+    include_paths = env.collector_values("::include_path")
```

Comment:
Would this work? This is different from options where this would be resolved with the current module and would only return one option.

Reply:
The resolver behaviour is generic and works the same for all node types. 
I've briefly considered allowing `env["::include_path"]` to return collector values (same with queries), however, I felt it clashed with the dict semantics of the options (raise exception on `env[option]`, return default on `env.get(option, default)`).

Calls to `env.collector(key)` and `env.query(key)` now raise exceptions when the key doesn't exists, since these are likely to not have a default that can fit into one-line argument.
Instead you should to use `env.has_collector(key)` and `env.has_query(key)` and do the default programmatically.

Also note that `env.collector_values(key, default=None)` raises an exception if the key is not found and default is `None`. I found this behavior very useful, since it prevents silent failures.
The return value is listified, so passing `default=None` would output `[None]`, so I chose to use this to check for the key. You can pass `default=[]` or `default=[None]` if you do not want that exception.

Reply:
also deprecated `env.query(key, default)` if default is not None.